### DISCUSSION TITLE
Installation Issue

### DISCUSSION THREAD

**Discussion by af4v** [wavelog/wavelog#847] [Q&A]
Hello all, I seem to have hit a roadblock with installing Wavelog. I'm trying to install on a Raspberry Pi Zero W running Raspbian Lite, however I encountered the same issued running DietPi. I followed the installation instructions to the letter, all my prechecks check out, my database is configured correctly, but the installer throws an error immediately after starting, and I cannot figure out why.
![Screenshot 2024-08-30 at 10 48 41 PM](https://github.com/user-attachments/assets/a840836d-54ec-4e8f-b563-197aa0106930)
![Screenshot 2024-08-30 at 10 49 20 PM](https://github.com/user-attachments/assets/9b25b83b-8020-42d3-bacd-3db5e0dd5ce2)
![Screenshot 2024-08-30 at 10 50 24 PM](https://github.com/user-attachments/assets/72508f26-c5b9-4e34-bbf8-1058d0d5f437)

**Comment by phl0** [comment]
The application/config directory is probably not writable for the webserver user.

**Reply by HB9HIL** [reply]
This is part of the checks

**Reply by af4v** [reply]
Just to make sure I set all directory permissions to 777. Still gave me the error

**Comment by HB9HIL** [comment]
Did you create manually the config.php before you started the installer? (You should not do that)

**Reply by af4v** [reply]
I didn't manually create any files before running the installer. I'm not incredibly experienced with Linux, databases, and such, so I stuck pretty close to the instructions in the wiki

**Comment by HB9HIL** [comment]
You can further check the error log of the webserver. 

for example /var/log/apache2/error.log

to reset a installation (for testing)

- delete application/config/config.php
- delete application/config/database.php
- delete install/.lock
- drop the database

**Reply by af4v** [reply]
I just dropped the wavelog database, but the wavelog installer still shows that there was a successful connection test with that database. That doesn't seem right to me?

**Reply by af4v** [reply]
Otherwise, none of those other files existed in the system, so I was unable to delete them

**Reply by af4v** [reply]
Also, here is the output of the apache2 error log. There are over 50 entries, but they all appear to be duplicates as if it's encountering the same error repeatedly
![Screenshot_20240831_211113_ConnectBot.jpg](https://github.com/user-attachments/assets/0f4cc002-67b7-4e71-bfb4-e1e843c78e3c)

**Reply by HB9HIL** [reply]
> I just dropped the wavelog database, but the wavelog installer still shows that there was a successful connection test with that database. That doesn't seem right to me?

this is okay. The database gets created if not existing and the entered user has permission to create a database

**Reply by HB9HIL** [reply]
> Also, here is the output of the apache2 error log. There are over 50 entries, but they all appear to be duplicates as if it's encountering the same error repeatedly
> ![Screenshot_20240831_211113_ConnectBot.jpg](https://github.com/user-attachments/assets/0f4cc002-67b7-4e71-bfb4-e1e843c78e3c)
> 
> 

I can't spot any obvious problems here.

Another question: Did you fill out every field in the user form input?

**Reply by af4v** [reply]
You're talking about first user creation wizard? This configuration still throws the installation error
![Screenshot_20240831_215726_Chrome.jpg](https://github.com/user-attachments/assets/d6a5d78f-8d7c-40b7-b8ea-1aacf7eddf5d)

**Reply by HB9HIL** [reply]
can you also show me your third tab? I will try to reproduce with the same input as yours

**Reply by af4v** [reply]
Here you are
![Screenshot_20240901_075354_Chrome.jpg](https://github.com/user-attachments/assets/da5e3819-4ee7-4858-9340-6b977f21a566)

**Comment by HB9HIL** [comment]
Hei @af4v 

I found some time to try to reproduce your issue. I currently testing a DietPi VM and I wasn't able to reproduce your issue. 

<img width="672" alt="image" src="https://github.com/user-attachments/assets/d0701120-b3cf-42c2-8f93-23fb64cf65bf">


but another question. 

I assume that you followed this Wiki: https://github.com/wavelog/Wavelog/wiki/Installation-on-a-Raspberry-Pi-using-Diet-Pi-Distro

Did you finally set the correct permissions like this?: https://github.com/wavelog/Wavelog/wiki/Installation#3-set-directory-ownership-and-permissions
In your case:

```bash
directory=/var/www/wavelog
sudo chown -R www-data:www-data $directory
sudo find $directory -type d -exec chmod 755 {} \;
sudo find $directory -type f -exec chmod 664 {} \;
```

**Reply by af4v** [reply]
Huh, very interesting. Yes I did set the database permissions using that string of commands

**Reply by af4v** [reply]
I will format the SD card with a DietPi image again and see if something changes

**Reply by HB9HIL** [reply]
These permissions are not for the database. They are for the webroot folder. Try again and tell me if it worked

**Reply by af4v** [reply]
> These permissions are not for the database. They are for the webroot folder. Try again and tell me if it worked

My mistake. When I get home from work I'll reestablish the "wavelog" database I dropped and I'll run these again

**Reply by HB9HIL** [reply]
After you run the permission command you should rerun the full installation after you made the steps I mentioned [here](https://github.com/wavelog/wavelog/discussions/847#discussioncomment-10509572)

**Reply by af4v** [reply]
Unfortunately that didn't work. I'm going to reimage the SD card with a fresh DietPi image and start completely over, and, if it would be helpful, I will keep track of every command I enter into the terminal

**Reply by af4v** [reply]
Alright, this is a complete log of everything I've done:

- download DietPi image for RPi 0W, burned to SD card with Raspberry Pi imager
- altered txt files to allow pi to auto-setup wifi on first run
- DietPi first boot up, runs automatic updates
- after auto-update, searched for LAMP and let the computer install automatically

The following is a list of commands entered into the terminal verbatim:
```
sudo apt install php-curl php-mysql php-mbstring php-xml php-zip
```
```
sudo apt install libapache2-mod-php
```
```
sudo apt-get install git
````
```
git clone https://github.com/wavelog/Wavelog.git /var/www/html
````
```
directory=/var/www/html
```
```
sudo chown -R www-data:www-data $directory
```
```
sudo find $directory -type d -exec chmod 755 {} \;
```
```
sudo find $directory -type f -exec chmod 664 {} \;
```
```
sudo mysql -u root -p
```
```
CREATE DATABASE wavelog;
```
```
CREATE USER 'af4v'@localhost IDENTIFIED BY '*********';
```
```
GRANT ALL PRIVILEGES ON wavelog.* TO 'af4v'@'localhost';
```
```
QUIT
```

I also installed vim to edit the php.ini file to edit max_execution_time, upload_max_filesize, and memory_limit to get them in the recommended ranges

**Reply by HB9HIL** [reply]
And after these steps it failed again?

**Reply by HB9HIL** [reply]
Can you tell me which PHP version was installed?

```bash
php -v
```

**Reply by af4v** [reply]
Correct, it still gives me the same error when trying to install. PHP is version 8.2.7

**Reply by af4v** [reply]
I have continued trying to run this install on different Pi Zeros with different SD cards, and they all still give me this error. I feel like I have to be missing something incredibly simple and obvious

**Reply by HB9HIL** [reply]
<img width="1253" alt="image" src="https://github.com/user-attachments/assets/a497c1c7-8ef5-44c4-89ca-bde301caf7fc">

I tried again on the VM (i don't have a zero here atm) with more realistic system specs (512GB RAM, etc) and with the exact commands you provided. But again it worked as expected on my side. What I noticed in the fancy `dietpi-software` thingy is that it already add some settings to PHP which seems fine to me in the first glance. What you have to remember is that a raspberry pi zero is not really strong enough to run Wavelog. Maybe in the beginning you will have some success (if we find out the installer issue) but when you have several thousand QSO's your instance will be run out of memory real quick. 

in the next step I tried to analyse the code:

The installer runs an ajax call on a trigger function in `install/includes/interface_assets/triggers.php`

```php
// config_file()
if ($_POST['run_config_file'] ?? false == true) {
	sleep(1);
	$data = $_POST['data'];
	if ($core->validate_post($data)) {
		if($core->write_configfile($data)) {
			$result = 'success';
		} else {
			$result = 'error';
		}
	} else {
		$result = 'error';
	}
	echo $result;
	exit;
}
```
As you can see there are two cases where the config process will fail. 
Either if the entered `$data` (callbook, userform, etc) could not be validated or
if the write function `$core->write_configfile($data)` failed. 

I would assume you try to break this down in little steps: 

Try to install **without** any callbook configured (Tab 3 - remove your callsign and the password in global callbook lookup)
Try to choose a really simple password in Tab 5. Something like "pasword1234"
We had some issues with passwords and special characters in the past. Maybe something is fishy here. 

> I also installed vim to edit the php.ini file to edit max_execution_time, upload_max_filesize, and memory_limit to get them in the recommended ranges

During my test I DID NOT edit the php.ini. I left everything on default. Maybe you can try the same.

**Reply by HB9HIL** [reply]
PS: I'm also interested to find this issue, just like you. We always want to make the user xperience as simple as possible. So let's find this issue together. If we do not find a solution here I would assume you to write me an email hb9hil@wavelog.org. So we can try to break this down directly on your machine.

**Reply by HB9HIL** [reply]
PS2: Which browser do you use??

**Reply by af4v** [reply]
> What you have to remember is that a raspberry pi zero is not really strong enough to run Wavelog. Maybe in the beginning you will have some success (if we find out the installer issue) but when you have several thousand QSO's your instance will be run out of memory real quick.
  
Yeah I assumed the Pi Zero probably wouldn't be a sufficient long term solution, but I had a spare one on hand and figured I would try it out, and if I liked the user experience I would upgrade the server to something a little more robust.

  
  
> Try to install without any callbook configured (Tab 3 - remove your callsign and the password in global callbook lookup)
Try to choose a really simple password in Tab 5. Something like "pasword1234"
We had some issues with passwords and special characters in the past. Maybe something is fishy here.
  
Both of these steps resulted in the same issue. Leaving php.ini at its default values also resulted in the same error.
  
  
    
>  Which browser do you use??
  
I'm using Google Chrome
 
  
   
> I'm also interested to find this issue, just like you. We always want to make the user xperience as simple as possible. So let's find this issue together.
  
I appreciate that very much! At this point I'm invested in finding an answer, even if just for the simple fact that it will help you with the development of the project

**Reply by phl0** [reply]
If it helps I could offer support via SSH / Remote Desktop. Maybe as last resort.

**Reply by erangaj** [reply]
Were you able to resolve the issue?

I am experiencing the same problem.

**Reply by erangaj** [reply]
> Were you able to resolve the issue?
> 
> I am experiencing the same problem.

I am using docker. For me the issues was the application/config/docker directory. It was missing.

**Reply by HB9HIL** [reply]
> > Were you able to resolve the issue?
> > I am experiencing the same problem.
> 
> I am using docker. For me the issues was the application/config/docker directory. It was missing.

This is really strange and definitly should not happen. Which version did you try to install?

**Comment by HB9HIL** [comment]
I tested the installer in various configurations, different browsers and OS. I wasn't able to spot any issues or failures.... This may is frustrating. I understand that. But at this point I don't know whats wrong on your side..

**Comment by Futureman83** [comment]
I had the same issue, the Problem was in my webserver Config:
Is Wrote this:
upload_max_filesize = 256
but it needs to be
upload_max_filesize = 256M

I forgot the  M....

**Comment by redradar** [comment]
I am having a nearly identical issue, but I have tried to install wavelog via Docker.    My enviroment is a desktop-less Debian 12 install with only Docker present,  It is serving to host the application.  I am accessing the web application from a seperate computer.  I followed the docker installation instructions at the following [wiki](https://github.com/wavelog/wavelog/wiki/Installation-via-Docker)  and everything went as seen in af4v screenshots.    I am failing at the config.php write.   I used the Yaml file in the wiki. 

I checked inside /var/lib/docker/overlay2  and can see multiple containers but the permissions look to be correct, but I am new to docker so I could be missing something.  Any pointers? 

![image](https://github.com/user-attachments/assets/a9afc5ff-f18c-42a1-aa74-0ad221fa6531)

**Reply by HB9HIL** [reply]
Hello @redradar 

What did you enter to the "configuration" tab?

I see double trailing slahes here, which can cause an issue:
<img width="124" alt="image" src="https://github.com/user-attachments/assets/68833c57-be0a-4c1d-97e3-9719471f967b">

**Reply by HB9HIL** [reply]
In addition: Did you leave any fields empty in the 5th tab? (The user creation tab)

**Reply by redradar** [reply]
My Third tab:
![image](https://github.com/user-attachments/assets/58f7025d-b646-43b0-8424-fd5cce0751f9)

My Fifth Tab: I left no user fields blank.

**Reply by HB9HIL** [reply]
You typed in `wavelog` for directory. This is wrong. In docker installations the directory is always blank. Try again without "wavelog" at directory and url

**Reply by redradar** [reply]
> You typed in `wavelog` for directory. This is wrong. In docker installations the directory is always blank. Try again without "wavelog" at directory and url

This has resolved my issue.    Thank you very much for your response.  I apologize if I missed any documentation that stated the points above.

**Reply by HB9HIL** [reply]
You are welcome :)

**Comment by f4inu** [comment]
I'm also affected by the same issue. All pre-checks are green, 

I tried to replace both the mysql password and user password to some very simple ones to prevent any character from causing issues, but this didn't solved anything.

I also tried to copy manually, as www-data user, the config.php file from install/config/ to application/config, this worked, and install failed just after when trying to copy database.php.

Installation was first attempted in last august, failed, tried again today with a fresh git pull. 

I'm using a very simple and classical lamp setup, which works for other software. If that matters, it runs on aarch64, like rpi3+.

**Reply by HB9HIL** [reply]
Hello F4INU

Can you show a Screenshot of all tabs in the installer?

**Reply by HB9HIL** [reply]
And share the messages from the installer debug log.

**Reply by f4inu** [reply]
Hello HB9HIL, here you are:

![wavelog_2](https://github.com/user-attachments/assets/ffb14d5c-9833-4993-be1f-217ef5adbf41)
![wavelog_3](https://github.com/user-attachments/assets/a3f1a604-ea18-4459-9c5f-f48ea9cf70c7)
![wavelog_4](https://github.com/user-attachments/assets/de7ad6e1-1b18-4d95-9db8-805c9e4c9700)
![wavelog_5](https://github.com/user-attachments/assets/14a615c3-534d-430b-8bc1-faec45bf30bf)
![wavelog_6-1](https://github.com/user-attachments/assets/e001a870-f772-43aa-952c-67d672f19a74)
![wavelog_6-2](https://github.com/user-attachments/assets/f37e274b-f34f-4c00-b9da-08c3a33978c6)

Logs are the following:

```
INFO - 2024-12-03 16:04:38 --> Called Installer index.php
INFO - 2024-12-03 16:04:38 --> With URL: https://<my url>/wavelog/install//
INFO - 2024-12-03 16:04:38 --> From IP: <my ip>
INFO - 2024-12-03 16:04:38 --> Set language cookie to "French"
DEBUG - 2024-12-03 16:07:57 --> Start writing config.php
ERROR - 2024-12-03 16:07:58 --> File: Could not write file. Check Permissions.
```

I tried to add various `log_message('error', 'line XX')` to `install/includes/core/core_class.php`, hopefully to try to catch where it was failing, without success.

**Reply by HB9HIL** [reply]
I'll try my very best to help you debugging this. 

1. For every new attempt please make sure you deleted the following files:
- `install/.lock`
- `application/config/config.php`
- `application/config/database.php`
- `application/logs/installer_debug.log`

Also make sure you dropped the database in mysql
```sql
DROP DATABASE wavelog;
```

You don't need to recreate the database, as this will be done by the installer. 

2. Then replace the file `install/includes/core/core_class.php` with this file here

https://raw.githubusercontent.com/HB9HIL/wavelog/refs/heads/installer_better_debug/install/includes/core/core_class.php

---

Run the installer again with the same values as before. Usually they are stored in the browser and you don't have to retype everything

Afterwards show us again the installer debug

`application/logs/installer_debug.log`

PS: It would be great if we can find that in the next 48 hours as we're planning to release soon and I'd be more then happy to find the bug here.

**Reply by f4inu** [reply]
Besides `installer_debug.log`, all other files were nonexistent. I deleted the log file, dropped the database and copied the new debug `core_class.php`, I have new log lines that might explain things! 

```
Wavelog Installer Debug Log
-------
Log Location: ../application/logs/installer_debug.log

DEBUG - 2024-12-03 20:40:42 --> Start writing config.php
ERROR - 2024-12-03 20:40:43 --> Data validation failed. <<< ? huh
ERROR - 2024-12-03 20:40:43 --> Directory wavelog does not exist.   <<< ?
ERROR - 2024-12-03 20:40:43 --> File: Could not write file. Check Permissions.
```
Why is it trying to find wavelog directory? Is it because of the path in step 3?

**Reply by HB9HIL** [reply]
How does your webserver configuration looks like? Do you use Apache?

Was the directory in tab 3 prefilled or did you add that?

**Reply by f4inu** [reply]
I'm using apache2 yes, and I don't remember for tab3, but I think this was pre-filled. 

Relevant Apache2 configuration: 
```
$ nl /etc/apache2/sites-enabled/mouse.synhacx.eu.org.conf

    50          Alias "/wavelog" "/var/www/wavelog"
    51          <Directory "/var/www/wavelog">
    52                  Allow from all
    53                  AllowOverride All
    54                  Require all granted
    55          </Directory>
```

I don't have any reverse proxy in front or any kind of redirect configuration in apache besides http to https.

**Reply by HB9HIL** [reply]
puh okay 

you can try the following to get this work with your type of installation.

Delete the original folder
``` bash
rm -r /var/www/wavelog
```

Create it complete new and switch into it
```bash
mkdir /var/www/wavelog
cd /var/www/wavelog
```

Clone the repository inside this folder in a folder called wavelog
```bash
git clone https://github.com/wavelog/wavelog wavelog
```

You will end up in a folder structure like
`/var/www/wavelog/wavelog/...`

This is not nice but this is the only way you can set it up like you want. 

Since we do a lot of filesystem magic with external data we need direct access to the root folder. Therefore a Alias does not work with Wavelog. 

I'd recommend to set up a subdomain

https://wavelog.yourserver.fr/ 

or do it like I said above

**Reply by HB9HIL** [reply]
Another thing you COULD TRY - with no guarantee that this work. I did not test that

Leave directory empty
but set the URL with directory

Like this:

<img width="484" alt="image" src="https://github.com/user-attachments/assets/2eff9a1f-5f68-4239-ba7a-0a1a898e3d73">

Tell me if this works

**Reply by f4inu** [reply]
I already tried to let the directory empty, install successfully copies config.php but fails at database.php. 

I also tried just right now to use a double wavelog directory structure, but this apparently also fails, maybe because of my apache configuration. 

I'll spin up a new wavelog dedicated vhost for that, not a big deal!

**Reply by HB9HIL** [reply]
I'm sorry. We can't cover each special apache/webserver configuration. We are focusing on the standard installation. Thx for your understanding.

**Reply by f4inu** [reply]
No need to be sorry, I'm already very grateful to have such a software, thanks a lot @HB9HIL !
